domain/application/modelmigration/import.go

```diff
@@ -223,31 +225,546 @@ func (i *importOperation) makeAddress(addr description.Address) (*network.SpaceA
+		return nil, fmt.Errorf("import manifest bases: %w", err)
+	}
+	return &internalcharm.Manifest{
+		Bases: bases,
+	}, nil
+}
+
+func (i *importOperation) importCharmLXDProfile(data description.CharmMetadata) (*internalcharm.LXDProfile, error) {
+	// LXDProfile is optional, so if we don't have any data, we can just return
+	// nil. If it does exist, then it's JSON encoded blob that we need to
+	// unmarshal into the internalcharm.LXDProfile struct.
+
+	lxdProfile := data.LXDProfile()
+	if lxdProfile == "" {
+		return nil, nil
+	}
+
+	var profile internalcharm.LXDProfile
+	if err := json.Unmarshal([]byte(lxdProfile), &profile); err != nil {
+		return nil, fmt.Errorf("unmarshal lxd profile: %w", err)
+	}
+
+	return &profile, nil
+}
+
+func (i *importOperation) importCharmConfig(data description.CharmConfigs) (*internalcharm.Config, error) {
+	// Charm config is optional, so if we don't have any data, we can just
+	// return nil.
+	if data == nil {
+		return nil, nil
+	}
+
+	descriptionConfig := data.Configs()
+
+	options := make(map[string]internalcharm.Option, len(descriptionConfig))
+	for name, c := range descriptionConfig {
+		options[name] = internalcharm.Option{
+			Type:        c.Type(),
+			Description: c.Description(),
+			Default:     c.Default(),
+		}
+	}
+
+	return &internalcharm.Config{
+		Options: options,
+	}, nil
+
+}
+
+func (i *importOperation) importCharmActions(data description.CharmActions) (*internalcharm.Actions, error) {
+	// Charm actions is optional, so if we don't have any data, we can just
+	// return nil.
+	if data == nil {
+		return nil, nil
+	}
+
+	descriptionActions := data.Actions()
+
+	actions := make(map[string]internalcharm.ActionSpec, len(descriptionActions))
+	for name, a := range descriptionActions {
+		parameters, err := importCharmParameters(a.Parameters())
+		if err != nil {
+			return nil, fmt.Errorf("import charm parameters: %w", err)
+		}
+
+		actions[name] = internalcharm.ActionSpec{
+			Description:    a.Description(),
+			Parallel:       a.Parallel(),
+			ExecutionGroup: a.ExecutionGroup(),
+			Params:         parameters,
+		}
 	}
+
+	return &internalcharm.Actions{
+		ActionSpecs: actions,
+	}, nil
 }
 
-func (s stubCharm) Actions() *internalcharm.Actions {
-	return &internalcharm.Actions{}
+func importCharmUser(data description.CharmMetadata) (internalcharm.RunAs, error) {
+	switch data.RunAs() {
+	case "default", "":
+		return internalcharm.RunAsDefault, nil
+	case "root":
+		return internalcharm.RunAsRoot, nil
+	case "sudoer":
+		return internalcharm.RunAsSudoer, nil
+	case "non-root":
+		return internalcharm.RunAsNonRoot, nil
+	default:
+		return internalcharm.RunAsDefault, fmt.Errorf("unknown run-as value %q: %w", data.RunAs(), errors.NotValid)
+	}
 }
 
-func (s stubCharm) Config() *internalcharm.Config {
-	return &internalcharm.Config{}
+func importAssumes(data string) (*assumes.ExpressionTree, error) {
+	// Assumes i§§s a recursive structure, rather than sending all that data over
```

Comment:
Looks like the ai is barfing up some unicode.

Reply:
Noice.

Reply:
This is the price of my new keyboard not having a dedicated escape key and me trying to find a new location for it. Clearly, that attempt (it has since moved) didn't stick in my muscle memory.